{
    "assignee": null,
    "assignees": [],
    "author_association": "CONTRIBUTOR",
    "body": "I seem to have hit a different issue with modulo on windows versus linux.\r\n\r\nif i cast the number to a float, or just put it inline before the modulo, it works as i expected.\r\n\r\n```python\r\nnum = 3316_000_000_000\r\nprint(num % 999)\r\n\r\nnum = float(3316_000_000_000)\r\nprint(int(num % 999))\r\n\r\nprint(3316_000_000_000 % 999)\r\n```\r\n\r\n```python\r\nfrom distutils.core import setup\r\nfrom Cython.Build import cythonize\r\n\r\nsetup(\r\n    ext_modules=cythonize(['test.pyx'])\r\n)\r\n```\r\n\r\n```bash\r\npyinstaller@3d46c498e294:/app$ python setup.py build_ext --inplace\r\nCompiling test.pyx because it changed.\r\n[1/1] Cythonizing test.pyx\r\nrunning build_ext\r\nbuilding 'test' extension\r\nC:\\mingw64\\bin\\gcc.exe -mdll -O -Wall -IC:\\Python36\\include -IC:\\Python36\\include -c test.c -o build\\temp.win-amd64-3.6\\Release\\test.o\r\nwriting build\\temp.win-amd64-3.6\\Release\\test.cp36-win_amd64.def\r\nC:\\mingw64\\bin\\gcc.exe -shared -s build\\temp.win-amd64-3.6\\Release\\test.o build\\temp.win-amd64-3.6\\Release\\test.cp36-win_amd64.def -LC:\\Python36\\libs -LC:\\Python36\\PCbuild\\amd64 -lpython36 -o Z:\\app\\test.cp36-win_amd64.pyd\r\npyinstaller@3d46c498e294:/app$ python\r\nPython 3.6.4 (v3.6.4:d48eceb, Dec 19 2017, 06:54:40) [MSC v.1900 64 bit (AMD64)] on win32\r\nType \"help\", \"copyright\", \"credits\" or \"license\" for more information.\r\n>>> import test\r\n935\r\n319\r\n319\r\n```\r\n\r\nAs you can see i'm building using mingw64, so this might be a factor, but the fact that it works either casting as float or inlining makes me think it's a cython bug.\r\n\r\n```bash\r\ngabriel@dev-desk-1:/tmp> python3.6 setup.py build_ext --inplace\r\nrunning build_ext\r\nbuilding 'test' extension\r\nx86_64-linux-gnu-gcc -pthread -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -g -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC -I/usr/include/python3.6m -I/home/gabriel/.virtualenvs/kivy36/include/python3.6m -c test.c -o build/temp.linux-x86_64-3.6/test.o\r\nx86_64-linux-gnu-gcc -pthread -shared -Wl,-O1 -Wl,-Bsymbolic-functions -Wl,-Bsymbolic-functions -Wl,-z,relro -Wl,-Bsymbolic-functions -Wl,-z,relro -g -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 build/temp.linux-x86_64-3.6/test.o -o /tmp/test.cpython-36m-x86_64-linux-gnu.so\r\n18:54:51 30/03/18 red argv[1] 94%kivy36 390                                       \r\ngabriel@dev-desk-1:/tmp> python3.6 \r\nPython 3.6.3 (default, Oct  6 2017, 08:44:35) \r\n[GCC 5.4.0 20160609] on linux\r\nType \"help\", \"copyright\", \"credits\" or \"license\" for more information.\r\n>>> import test\r\n319\r\n319\r\n319\r\n```\r\n\r\nDoing these operation in python directly always results in 319, on both platform.\r\n\r\nthe docker image used to build can be found on docker hub (https://hub.docker.com/r/tshirtman/kpmwine/)\r\n\r\n```\r\ndocker run -ti -v $PWD:/app -v /tmp/pipcache:\"/wine/drive_c/users/pyinstaller/Local Settings/application data/pip/cache/\" tshirtman/kpmwine bash\r\n```",
    "closed_at": "2018-04-08T12:17:18Z",
    "comment_data": [
        {
            "author_association": "MEMBER",
            "body": "I obviously cannot reproduce this without Windows, but I added your example to get it tested on appveyor. Are you using the latest Cython 0.28.1 release?\r\n\r\nAs for potentially relevant platform differences, `long` is 32 bits on Windows64 and 64 bits on Linux64, and I'm not sure what the size of a `PyLong` digit is on Windows.\r\n\r\nIn any case, the code that Cython generates is correctly typed and analysed. I don't see anything obviously wrong with it. The result is calculated in the generated helper function `__Pyx_PyInt_RemainderObjC()`, specifically by the code that handles `PyLong` values.",
            "created_at": "2018-03-30T17:47:04Z",
            "html_url": "https://github.com/cython/cython/issues/2192#issuecomment-377580691",
            "id": 377580691,
            "issue_url": "https://api.github.com/repos/cython/cython/issues/2192",
            "node_id": "MDEyOklzc3VlQ29tbWVudDM3NzU4MDY5MQ==",
            "updated_at": "2018-03-30T17:47:04Z",
            "url": "https://api.github.com/repos/cython/cython/issues/comments/377580691",
            "user": {
                "avatar_url": "https://avatars0.githubusercontent.com/u/491659?v=4",
                "events_url": "https://api.github.com/users/scoder/events{/privacy}",
                "followers_url": "https://api.github.com/users/scoder/followers",
                "following_url": "https://api.github.com/users/scoder/following{/other_user}",
                "gists_url": "https://api.github.com/users/scoder/gists{/gist_id}",
                "gravatar_id": "",
                "html_url": "https://github.com/scoder",
                "id": 491659,
                "login": "scoder",
                "node_id": "MDQ6VXNlcjQ5MTY1OQ==",
                "organizations_url": "https://api.github.com/users/scoder/orgs",
                "received_events_url": "https://api.github.com/users/scoder/received_events",
                "repos_url": "https://api.github.com/users/scoder/repos",
                "site_admin": false,
                "starred_url": "https://api.github.com/users/scoder/starred{/owner}{/repo}",
                "subscriptions_url": "https://api.github.com/users/scoder/subscriptions",
                "type": "User",
                "url": "https://api.github.com/users/scoder"
            }
        },
        {
            "author_association": "MEMBER",
            "body": "Appveyor seems quite happy with the test:\r\nhttps://ci.appveyor.com/project/cython/cython/build/master-1898",
            "created_at": "2018-03-31T07:47:48Z",
            "html_url": "https://github.com/cython/cython/issues/2192#issuecomment-377674366",
            "id": 377674366,
            "issue_url": "https://api.github.com/repos/cython/cython/issues/2192",
            "node_id": "MDEyOklzc3VlQ29tbWVudDM3NzY3NDM2Ng==",
            "updated_at": "2018-03-31T07:47:48Z",
            "url": "https://api.github.com/repos/cython/cython/issues/comments/377674366",
            "user": {
                "avatar_url": "https://avatars0.githubusercontent.com/u/491659?v=4",
                "events_url": "https://api.github.com/users/scoder/events{/privacy}",
                "followers_url": "https://api.github.com/users/scoder/followers",
                "following_url": "https://api.github.com/users/scoder/following{/other_user}",
                "gists_url": "https://api.github.com/users/scoder/gists{/gist_id}",
                "gravatar_id": "",
                "html_url": "https://github.com/scoder",
                "id": 491659,
                "login": "scoder",
                "node_id": "MDQ6VXNlcjQ5MTY1OQ==",
                "organizations_url": "https://api.github.com/users/scoder/orgs",
                "received_events_url": "https://api.github.com/users/scoder/received_events",
                "repos_url": "https://api.github.com/users/scoder/repos",
                "site_admin": false,
                "starred_url": "https://api.github.com/users/scoder/starred{/owner}{/repo}",
                "subscriptions_url": "https://api.github.com/users/scoder/subscriptions",
                "type": "User",
                "url": "https://api.github.com/users/scoder"
            }
        },
        {
            "author_association": "CONTRIBUTOR",
            "body": "Yes i was using the latest from pip, 0.28.1, i did integrate your test in my code to be sure there is no subtle difference, and it does generate the bug on my side, so i guess next step for me is to setup a windows build environment with visual studio and see if i can reproduce here.\r\n\r\nThanks for looking into it :).",
            "created_at": "2018-03-31T14:03:07Z",
            "html_url": "https://github.com/cython/cython/issues/2192#issuecomment-377695251",
            "id": 377695251,
            "issue_url": "https://api.github.com/repos/cython/cython/issues/2192",
            "node_id": "MDEyOklzc3VlQ29tbWVudDM3NzY5NTI1MQ==",
            "updated_at": "2018-03-31T14:03:07Z",
            "url": "https://api.github.com/repos/cython/cython/issues/comments/377695251",
            "user": {
                "avatar_url": "https://avatars1.githubusercontent.com/u/22759?v=4",
                "events_url": "https://api.github.com/users/tshirtman/events{/privacy}",
                "followers_url": "https://api.github.com/users/tshirtman/followers",
                "following_url": "https://api.github.com/users/tshirtman/following{/other_user}",
                "gists_url": "https://api.github.com/users/tshirtman/gists{/gist_id}",
                "gravatar_id": "",
                "html_url": "https://github.com/tshirtman",
                "id": 22759,
                "login": "tshirtman",
                "node_id": "MDQ6VXNlcjIyNzU5",
                "organizations_url": "https://api.github.com/users/tshirtman/orgs",
                "received_events_url": "https://api.github.com/users/tshirtman/received_events",
                "repos_url": "https://api.github.com/users/tshirtman/repos",
                "site_admin": false,
                "starred_url": "https://api.github.com/users/tshirtman/starred{/owner}{/repo}",
                "subscriptions_url": "https://api.github.com/users/tshirtman/subscriptions",
                "type": "User",
                "url": "https://api.github.com/users/tshirtman"
            }
        },
        {
            "author_association": "MEMBER",
            "body": "I'm able to reproduce this with your docker image. In your setup, MinGW finds this config:\r\n```\r\nsizeof(digit) = 2\r\nsizeof(long) = 4\r\nsizeof(long long) = 8\r\nPyLong_SHIFT = 15\r\nPyLong_BASE = 32768\r\n```\r\nHowever, this is clearly not what CPython uses in your docker image:\r\n```\r\n$ wine python -c 'import sys; print([sys.getsizeof(i) for i in (0, 1, 2**15-1, 2**15, 2**30-1, 2**30, 2**30+1, 2**45-1, 2**45, 2**60-1, 2**60)])'\r\n[24, 28, 28, 28, 28, 32, 32, 32, 32, 32, 36]\r\n```\r\nThe integer object size increases by 4 bytes every 2**30 integers, which suggests a `PyLong` digit size of 30 bits, not 15.\r\n\r\nI have no idea how to fix the setup here. If you need to use it, you may consider disabling direct `PyLong` usage in Cython by defining `-D CYTHON_USE_PYLONG_INTERNALS=0`. But that will slow things down a bit.\r\n\r\nI'll close this ticket since it's not a bug in Cython.",
            "created_at": "2018-04-08T12:17:18Z",
            "html_url": "https://github.com/cython/cython/issues/2192#issuecomment-379545946",
            "id": 379545946,
            "issue_url": "https://api.github.com/repos/cython/cython/issues/2192",
            "node_id": "MDEyOklzc3VlQ29tbWVudDM3OTU0NTk0Ng==",
            "updated_at": "2018-04-08T12:17:18Z",
            "url": "https://api.github.com/repos/cython/cython/issues/comments/379545946",
            "user": {
                "avatar_url": "https://avatars0.githubusercontent.com/u/491659?v=4",
                "events_url": "https://api.github.com/users/scoder/events{/privacy}",
                "followers_url": "https://api.github.com/users/scoder/followers",
                "following_url": "https://api.github.com/users/scoder/following{/other_user}",
                "gists_url": "https://api.github.com/users/scoder/gists{/gist_id}",
                "gravatar_id": "",
                "html_url": "https://github.com/scoder",
                "id": 491659,
                "login": "scoder",
                "node_id": "MDQ6VXNlcjQ5MTY1OQ==",
                "organizations_url": "https://api.github.com/users/scoder/orgs",
                "received_events_url": "https://api.github.com/users/scoder/received_events",
                "repos_url": "https://api.github.com/users/scoder/repos",
                "site_admin": false,
                "starred_url": "https://api.github.com/users/scoder/starred{/owner}{/repo}",
                "subscriptions_url": "https://api.github.com/users/scoder/subscriptions",
                "type": "User",
                "url": "https://api.github.com/users/scoder"
            }
        }
    ],
    "comments": 4,
    "comments_url": "https://api.github.com/repos/cython/cython/issues/2192/comments",
    "created_at": "2018-03-30T17:03:08Z",
    "event_data": [
        {
            "actor": {
                "avatar_url": "https://avatars0.githubusercontent.com/u/491659?v=4",
                "events_url": "https://api.github.com/users/scoder/events{/privacy}",
                "followers_url": "https://api.github.com/users/scoder/followers",
                "following_url": "https://api.github.com/users/scoder/following{/other_user}",
                "gists_url": "https://api.github.com/users/scoder/gists{/gist_id}",
                "gravatar_id": "",
                "html_url": "https://github.com/scoder",
                "id": 491659,
                "login": "scoder",
                "node_id": "MDQ6VXNlcjQ5MTY1OQ==",
                "organizations_url": "https://api.github.com/users/scoder/orgs",
                "received_events_url": "https://api.github.com/users/scoder/received_events",
                "repos_url": "https://api.github.com/users/scoder/repos",
                "site_admin": false,
                "starred_url": "https://api.github.com/users/scoder/starred{/owner}{/repo}",
                "subscriptions_url": "https://api.github.com/users/scoder/subscriptions",
                "type": "User",
                "url": "https://api.github.com/users/scoder"
            },
            "commit_id": "f9fddb047d59b1567144763c6ce1c8f22e13bba8",
            "commit_url": "https://api.github.com/repos/cython/cython/commits/f9fddb047d59b1567144763c6ce1c8f22e13bba8",
            "created_at": "2018-03-30T17:33:20Z",
            "event": "referenced",
            "id": 1550068978,
            "node_id": "MDE1OlJlZmVyZW5jZWRFdmVudDE1NTAwNjg5Nzg=",
            "url": "https://api.github.com/repos/cython/cython/issues/events/1550068978"
        },
        {
            "actor": {
                "avatar_url": "https://avatars0.githubusercontent.com/u/491659?v=4",
                "events_url": "https://api.github.com/users/scoder/events{/privacy}",
                "followers_url": "https://api.github.com/users/scoder/followers",
                "following_url": "https://api.github.com/users/scoder/following{/other_user}",
                "gists_url": "https://api.github.com/users/scoder/gists{/gist_id}",
                "gravatar_id": "",
                "html_url": "https://github.com/scoder",
                "id": 491659,
                "login": "scoder",
                "node_id": "MDQ6VXNlcjQ5MTY1OQ==",
                "organizations_url": "https://api.github.com/users/scoder/orgs",
                "received_events_url": "https://api.github.com/users/scoder/received_events",
                "repos_url": "https://api.github.com/users/scoder/repos",
                "site_admin": false,
                "starred_url": "https://api.github.com/users/scoder/starred{/owner}{/repo}",
                "subscriptions_url": "https://api.github.com/users/scoder/subscriptions",
                "type": "User",
                "url": "https://api.github.com/users/scoder"
            },
            "commit_id": null,
            "commit_url": null,
            "created_at": "2018-04-08T12:17:18Z",
            "event": "closed",
            "id": 1562580486,
            "node_id": "MDExOkNsb3NlZEV2ZW50MTU2MjU4MDQ4Ng==",
            "url": "https://api.github.com/repos/cython/cython/issues/events/1562580486"
        }
    ],
    "events_url": "https://api.github.com/repos/cython/cython/issues/2192/events",
    "html_url": "https://github.com/cython/cython/issues/2192",
    "id": 310116836,
    "labels": [],
    "labels_url": "https://api.github.com/repos/cython/cython/issues/2192/labels{/name}",
    "locked": false,
    "milestone": null,
    "node_id": "MDU6SXNzdWUzMTAxMTY4MzY=",
    "number": 2192,
    "repository_url": "https://api.github.com/repos/cython/cython",
    "state": "closed",
    "title": "different modulo behavior on windows vs linux since python3",
    "updated_at": "2018-04-08T12:17:18Z",
    "url": "https://api.github.com/repos/cython/cython/issues/2192",
    "user": {
        "avatar_url": "https://avatars1.githubusercontent.com/u/22759?v=4",
        "events_url": "https://api.github.com/users/tshirtman/events{/privacy}",
        "followers_url": "https://api.github.com/users/tshirtman/followers",
        "following_url": "https://api.github.com/users/tshirtman/following{/other_user}",
        "gists_url": "https://api.github.com/users/tshirtman/gists{/gist_id}",
        "gravatar_id": "",
        "html_url": "https://github.com/tshirtman",
        "id": 22759,
        "login": "tshirtman",
        "node_id": "MDQ6VXNlcjIyNzU5",
        "organizations_url": "https://api.github.com/users/tshirtman/orgs",
        "received_events_url": "https://api.github.com/users/tshirtman/received_events",
        "repos_url": "https://api.github.com/users/tshirtman/repos",
        "site_admin": false,
        "starred_url": "https://api.github.com/users/tshirtman/starred{/owner}{/repo}",
        "subscriptions_url": "https://api.github.com/users/tshirtman/subscriptions",
        "type": "User",
        "url": "https://api.github.com/users/tshirtman"
    }
}